### PR TITLE
we should have a config-based way to skip flaky tests

### PR DESCRIPTION
**Pull request by zdevito**
This particular approach queries our issue tracker for test titles that
match the following format:

```
DISABLED test_async_grad_guard_with_grad (jit.test_async.TestAsync)
```

And then skips the python test for them. There is 1 second timeout so
if the internet flakes we still run the test suite, without disabling any
tests.

This is intended as a quick fix, similar to ninja unland, to get to a green
master. Long term test disables should go into the code.
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29944 we should have a config-based way to skip flaky tests**

Differential Revision: [D18621773](https://our.internmc.facebook.com/intern/diff/D18621773)

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 54ecaf9:
* **1/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=54ecaf92a5cd164b6f6291ba84796a9ade410be4):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_linux_backward_compatibility_check_test | Test | `The PR is introducing backward incompatible changes` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Review by ezyang — CHANGES_REQUESTED** [review]
I'm concerned we will instantly hit GitHub rate limits as this is implemented. This is implemented as a query *per test* we run. That means we will issue X search queries to GitHub, where X is the number of tests in our test suite.  The unauthenticated GitHub search API permits *ten* requests per minute: https://developer.github.com/v3/search/#rate-limit

A better approach would for there to be a single endpoint which, when queried, returns the full set of disabled tests so we only need to query once. But I still don't really want to be subject to GitHub API rate limits. We could host this as a file in a separate GitHub repository, at which point we get to bypass GitHub API rate limits.

**Comment by zdevito** [comment]
ok, I switched this to clone a separate repo. A chronos job can run `update_disabled_tests` to generate the result at a rate that would not get throttled.

File: test/common_utils.py [hunk 1]
```diff
@@ -559,6 +563,42 @@ def __exit__(self, exec_type, exec_value, traceback):
...
+                    test_name = title[len(key):].strip()
+                    disabled_test_from_issues.add(test_name)
+
+        if not IS_SANDCASTLE:
+            try:
+                try:
+                    read_and_process()
+                except FileNotFoundError:
+                    repo_dir = tempfile.mkdtemp()
+                    check_call(['git', 'clone', 'https://github.com/zdevito/pytorch_disabled_tests.git', repo_dir])
```
Comment (ezyang) [inline test/common_utils.py]: I don't think there's any reason to clone the whole repo. Just fetch the file from https://raw.githubusercontent.com/zdevito/pytorch_disabled_tests/master/result.json

File: test/common_utils.py [hunk 2]
```diff
@@ -559,6 +563,42 @@ def __exit__(self, exec_type, exec_value, traceback):
...
+                    tmp_name = 'disabled_tests_{}.json'.format(uuid4())
+                    shutil.copyfile(os.path.join(repo_dir, 'result.json'), tmp_name)
+                    os.rename(tmp_name, 'disabled_tests.json')
+                    read_and_process()
+            except Exception:
+                print("Couldn't download test skip set, leaving all tests enabled...")
+
+    if test_name in disabled_test_from_issues:
+        raise unittest.SkipTest(
+            "Test is disabled because an issue exists disabling it: https://github.com/search?q={}".format(disabled_query))
```
Comment (ezyang) [inline test/common_utils.py]: Wouldn't it be better to just report the actual issue here? That way, we wouldn't need to reproduce the disabled query in this file (at risk of it getting out of date)

**Review by ezyang — APPROVED** [review]
It would be good to make the Chronos job discoverable somehow, so that people can manually trigger disabling of tests. Also it would be good if we taught our bot to comment on issues like this with information about what the test is doing, to help make it more clear what's going on.

**Comment by zdevito** [comment]
I guess if our bot understood the test disable issues it can be the thing that triggers the re-request of the query. Where does the bot code live?

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch-probot

**Comment by zdevito** [comment]
I updated this with your suggestions. I have a couple of caveats before I enable this.  For this to be a good dev experience, the disabled test queries should be running on users dev machines. Otherwise, we will disable a test on CI but someone will see it locally and think it is a real failure. To avoid cache-invalidation issues, I don't  save the result to a file, so each test suite (but not each individual test) will load the file. The request happens very fast (file is small, and it is not causing a query to run each time).

**Pull request closed without merging**